guardian/editions
# switch settings to Apollo

## Summary

State management has been a little bit problematic in our app. We have our own custom solutions for subscribing to state changes and update state, such as `cachedOrPromise` but also, in that PR's case, `useSettingsFromStore`; which this PR gets rid off completely. (As an example that we should avoid implementing ourself state management logic is that this function is buggy: it doesn't have a proper `useEffect` destructor.)

This PR is a proposal to switch to Apollo for local state management, and reimplement the settings using that framework. Originally Apollo is meant for GraphQL, but it turns out GraphQL also fits quite well local state. (The framework even allows for blending remote and local data, something useful if we ever want to fetch from GraphQL endpoints.)

What Apollo brings us is:

* We don't need to set-up our own subscription system. Apollo automatically updates any components that depends on a particular set of fields, via `useQuery`, when these fields change value. (For example, a setting.)
* We can fetch data (from local storage, RESTs API, filesystem…) required for an entire tree of component via a single query by putting fragments together. The fragments can live in specific sub components. Doing the query in a single place means there is only 1 re-render, when all the data is fetched/resolved.
* Apollo is an off-the-shelf solution very popular in the open source community, and covered with tests. This is a great way to avoid bugs. If there are bugs, we can always contribute back to Apollo.

Alternatives to Apollo are to use something like Redux. From my research, I got the sense would bring additional boilerplate code, because they don't provide too much out-of-the-box for managing async operations, or they don't do so in a way that's as simple as the GraphQL model. 

For synchronous state, we don't need Apollo: `useState` is a perfect solution for that.

If that PR is approved, then we can switch the rest of our async state accordingly, including what uses `cachedOrPromise`.

## Test Plan

These changes are not platform-specific so it can be tested only on one platform.

1. Uninstall the app if already here, so that we start from a blank slate.
2. Instal and start the app. Notice the onboarding notice. Press "I'm okay".
3. Home page loads. Weather shows up.
4. Go to settings. Switch "Display Weather" to off. Back to home page, notice weather is know hidden.
5. Go to Privacy Settings. Set Performance to off. Close the app. Open again, go the Privacy Settings, make sure Performance is still off. Make sure weather is still hidden.
6. Open an article. Swipe right and check everything renders correctly.
7. Go to Settings. Press "Version" tons of times, which enables the dev menu. Notice the settings list show up correctly.
